pkg/alertmanager: add URL validation for Mattermost receiver

varundeepsaini opened the pull request:
Relates to #8193

## Description

Adds URL validation for Mattermost receiver configuration fields when loaded from secrets. This ensures URLs are validated regardless of whether configurations come from CustomResources or secrets.

**Validated fields:**
- `webhook_url`
- `icon_url`
- `attachments[].author_link`
- `attachments[].author_icon`
- `attachments[].title_link`
- `attachments[].thumb_url`
- `attachments[].footer_icon`
- `attachments[].image_url`

## Type of change

- [x] Bug fix

## Verification

- [x] Added test cases for each URL field with invalid URLs
- [x] Added test case for valid URLs
- [x] Fixed existing tests that used invalid URLs (missing scheme)
- [x] All tests pass

## Changelog entry

```release-note
Add URL validation for Mattermost receiver secrets
```

Commit:
pkg/alertmanager: add URL validation for Mattermost receiver

nutmos commented on pkg/alertmanager/amcfg_test.go:
```diff
@@ -8105,6 +8105,264 @@ func TestSanitizeRocketChatConfig(t *testing.T) {
+		expectErr      bool
+	}{
+		{
+			name:           "mattermost_configs returns error for unsupported versions",
```
While I like to see more test cases implemented, test cases in this PR should be related to the feature implemented here. The rest should open a new PR.

nutmos commented on pkg/alertmanager/amcfg.go:
```diff
@@ -3306,6 +3306,51 @@ func (mc *mattermostConfig) sanitize(amVersion semver.Version, logger *slog.Logg
+			}
+		}
+		if attachment.TitleLink != "" {
+			if _, err := validation.ValidateURL(attachment.TitleLink); err != nil {
```
The title_link field is a template string.

varundeepsaini commented:
Rebased on main and addressed both review comments:
- template-capable fields (icon_url and all attachment URL fields incl. title_link) now use `validation.ValidateTemplateURL`; plain `ValidateURL` is kept for webhook_url only. Added test cases for templated URLs and invalid template syntax.
- removed the two version-gating test cases that were unrelated to this feature.

nutmos commented:
@varundeepsaini The newer version moves attachment fields to be under the top-level fields. For example, author_link. Would you mind to update them as well?

varundeepsaini commented:
@nutmos done — the top-level fields (author_link, author_icon, title_link, thumb_url, footer_icon, image_url) are now validated with `ValidateTemplateURL` as well. They're validated after the version-gating block, so on Alertmanager < 0.32.0 they're dropped with a warning (matching upstream) rather than rejected. Added test cases for an invalid top-level author_link and for valid/templated top-level URLs.
